File: src/main/scala/com/aikosolar/app/TimeLagWatermarkGenerator.scala
package com.aikosolar.app

import com.aikosolar.domain.DFTube
import org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks
import org.apache.flink.streaming.api.watermark.Watermark


class TimeLagWatermarkGenerator  extends AssignerWithPeriodicWatermarks[DFTube] {

  // 所有管式设备没有超过1小时的
  val maxTimeLag = 3600000L; // 1h

  override def extractTimestamp(element: DFTube, previousElementTimestamp: Long): Long = {
    element.timeSecond * 1000
  }

  override def getCurrentWatermark: Watermark = {
    // return the watermark as current time minus the maximum time lag
    new Watermark(System.currentTimeMillis() - maxTimeLag)
  }
}

File: df-extract-run-count-etl/src/main/java/com/aikosolar/bigdata/df/TimeLagWatermarkGenerator.java
package com.aikosolar.bigdata.df;


import org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks;
import org.apache.flink.streaming.api.watermark.Watermark;

import javax.annotation.Nullable;

/**
 * @author xiaowei.song
 */
public class TimeLagWatermarkGenerator implements AssignerWithPeriodicWatermarks<DFTube> {

    /**
     * 所有管式设备没有超过1小时的
     * 水位设置成3s, 3s之后开始计算窗口计算，之前的抛弃
     **/
    public static final Long MAX_TIME_LAG = 3000L;

    @Nullable
    @Override
    public Watermark getCurrentWatermark() {
        return new Watermark(System.currentTimeMillis() - MAX_TIME_LAG);
    }

    @Override
    public long extractTimestamp(DFTube e, long l) {
        return e.timeSecond * 1000;
    }
}
